cpg_infra/config/deserializabledataclass.py

```diff
@@ -30,94 +30,110 @@ def __post_init__(self):
 
         for fieldname, ftype in fields.items():
             value = self.__dict__.get(fieldname)
-
-            if not value:
-                continue
-            dtypes = []
-            # determine which type we should try to parse the value as
-            # handle unions (eg: None | DType)
-            if isinstance(ftype, UnionType):
-                is_already_correct_type = False
-                for dtype in get_args(ftype):
-                    if dtype and issubclass(dtype, DeserializableDataclass):
-                        # It's a DeserializableDataclass :)
-                        dtypes.append(dtype)
-                    elif dtype and isinstance(value, dtype):
-                        is_already_correct_type = True
-                if is_already_correct_type:
-                    continue
-
-            elif issubclass(ftype, DeserializableDataclass):
-                if isinstance(value, ftype):
-                    continue
-                dtypes.append(ftype)
-
-            e = None
-            # try to see if the value will parse as one of the detected DTypes
-            for dtype in dtypes:
-                if not isinstance(value, dict):
-                    raise ValueError(
-                        f'Expected {value} to be a dictionary to parse, got {type(value)}.'
-                    )
-                try:
-                    self.__dict__[fieldname] = dtype(**value)
-                    e = None
-                    break
-                except TypeError as exc:
-                    e = exc
-
-            if e:
-                raise e
+            try:
+                self.__dict__[fieldname] = try_parse_value_as_type(value, ftype)
+            except ValueError as e:
+                raise ValueError(
+                    f'Error parsing {self.__class__.__name__}.{fieldname} :: {value!r}'
+                ) from e
 
     def __repr__(self):
         args = ', '.join(f'{k}={v!r}' for k, v in vars(self).items())
         return f'{self.__class__.__name__}( {args} )'
 
 
-def parse_value_from_type(config, fieldname, ftype):
-    if ftype is None:
-        return None
-
-    if ftype in (list, dict) or get_origin(ftype) in (list, dict):
-        ftype_type = ftype if ftype in (list, dict) else get_origin(ftype)
-        value = config.get_object(fieldname)
+def try_parse_value_as_type(value, dtype):
+    """
+    Try to parse a value as a specific type, if it fails
+    return None
```

Comment:
Doc string might be misleading since an error will be raised for values that cannot be validated.

Reply:
Nice catch, clarified